### PR TITLE
gps: replace GeoCoord::latLongToMeter's spherical trig with equirectangular approximation

### PR DESCRIPTION
**Pull request by ndoo**
## What

`GeoCoord::latLongToMeter()` computed great-circle distance via the spherical law of cosines (the comment mislabeled it "Haversine formula") - 4 `cos()`, 4 `sin()`, and 1 `acos()` call. Disassembly on `wio-e5` (`objdump`/`nm`) confirmed these were the only call sites in the firmware for that trig chain, pulling in ~4.4KB of internal range-reduction/kernel helpers (`__kernel_rem_pio2`, `__ieee754_rem_pio2`, `__kernel_cos`, `__kernel_sin`) for a function every caller uses for display or movement-threshold checks - never routing or security decisions.

Replaces it with an equirectangular (flat-plane) projection, using a minimax polynomial approximation of `cos(latitude)` ("cos_52", https://www.ganssle.com/approx.htm) instead of a live `cos()` call, plus `sqrt()`. No trig library calls.

## Bug fixed during development

Raw longitude subtraction doesn't handle the antimeridian. Two points ~22km apart straddling the 180th meridian (e.g. 179.9 and -179.9) computed as ~40,000km apart before adding wraparound handling (`dLng` normalized to `[-pi, pi]`). The original trig-based formula didn't have this problem, since `cos()`/`sin()` are periodic - a real behavioral difference to be aware of when porting spherical-trig code to a flat-plane approximation.

## Accuracy

Measured against the original formula:

| Case | True distance | Approximation | Absolute error | Relative error |
| --- | --- | --- | --- | --- |
| Movement threshold (67m, London) | 66.7 m | 66.7 m | ~0 m | 0.00% |
| Movement threshold (540m, London) | 539.4 m | 539.4 m | ~0 m | 0.00% |
| London to Paris | 343.3 km | 343.3 km | +47 m | 0.01% |
| NYC to Boston | 305.9 km | 305.9 km | +34 m | 0.01% |
| Tokyo to Osaka | 392.1 km | 392.2 km | +40 m | 0.01% |
| Sydney to Melbourne | 712.9 km | 713.2 km | +304 m | 0.04% |
| Antimeridian crossing (near equator) | 22.2 km | 22.2 km | ~0 m | 0.00% |
| Pole to pole (antipodal) | 20,000 km | 20,000 km | ~0 m | 0.00% |
| Near-pole, 180 deg longitude apart (within 0.1 deg of pole) | 22.2 km | 34.8 km | +12.6 km | 56.5% |

The first seven rows cover realistic use (movement thresholds, city-to-city distances, LoRa-range checks) and stay under 0.05% error. The last two rows are the test suite's extreme-latitude cases (`test_no_nan_at_extreme_latitudes`): real coordinate values, but degenerate for a flat-plane approximation, since a fixed longitude difference means very little distance near the pole and a lot away from it. Those are checked for NaN-safety only, not accuracy - no realistic Meshtastic deployment operates within 0.1 degrees of a pole.

No prior test coverage existed for `latLongToMeter`. New `test_geocoord_distance` suite checks accuracy bounds against a reference implementation of the original formula (1% local, 4% regional), the antimeridian fix, distance symmetry, and NaN-safety at extreme latitudes.

## Flash-size impact

| Board | Chip | Result |
| --- | --- | --- |
| `wio-e5` | STM32WLE5 | **-5,792 bytes** (233,744 -> 227,952 B, 94.3% -> 92.0% of the 247,808 B budget) |
| `seeed_wio_tracker_L1` | nRF52840 | **-1,104 bytes** (751,192 -> 750,088 B, 92.2% -> 92.0% of the 815,104 B budget) |
| `heltec-v3` | ESP32-S3 | **-1,024 bytes** (2,297,247 -> 2,296,223 B, both 68.7% of the 3,342,336 B budget) |
| `heltec-v4` | ESP32-S3 | **-1,048 bytes** (2,313,643 -> 2,312,595 B, both 35.3% of the 6,553,600 B budget) |

Confirmed via `nm` on the `wio-e5` ELF that `acos`/`cos`/`sin` and their internal helpers are absent after this change. The `wio-e5` saving exceeds the ~4.4KB directly attributable to the trig chain itself - likely from also removing the `isnan()` branch and simplifying from 9 trig calls down to one polynomial evaluation plus one `sqrt()` call.

Measured by checking out the pre-change `GeoCoord.cpp` via `git checkout <parent-commit> -- src/gps/GeoCoord.cpp`, building, then restoring the fix and rebuilding immediately after, with no other build in between, to avoid the ~20KB of build-to-build noise this machine sees on ESP32-S3 from ESP-IDF/Xtensa LTO non-determinism (flagged separately in #11179). Smaller absolute savings than `wio-e5`/nRF52840 - Xtensa's libm has a much smaller `cos`/`sin`/`acos` footprint than ARM's `newlib-nano`, so there's less to remove.

## 🤝 Attestations

- [x] I have tested that my proposed changes behave as described.
- [ ] I have tested that my proposed changes do not cause any obvious regressions on the following devices:
  - [ ] Heltec (Lora32) V3
  - [ ] LilyGo T-Deck
  - [ ] LilyGo T-Beam
  - [ ] RAK WisBlock 4631
  - [ ] Seeed Studio T-1000E tracker card
  - [ ] Other (please specify below)

Tested via the `test_geocoord_distance` native unit test suite and flash-size measurement on two boards, above. Build-verified only. Changes the accuracy (not the correctness/availability) of an on-screen distance display and a couple of movement-threshold checks - touches several screen/UI call sites (`Screen.cpp`, `UIRenderer.cpp`, multiple InkHUD applets) not exercised in this testing.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved latitude/longitude distance calculations for more consistent movement and display thresholds.
  * Enhanced accuracy for routes crossing the international date line.
  * Prevented invalid distance results at extreme latitudes.
  * Improved handling of identical locations and other edge-case coordinates.

* **Tests**
  * Added coverage for zero-distance, local, regional, symmetric, date-line, and extreme-coordinate calculations.
  * Added validation against trusted distance calculations across varied geographic scenarios.

### CODE REVIEW

**Comment by ndoo** [comment]
@coderabbitai review

**Comment by ndoo** [comment]
Trimmed all five flagged comment blocks (GeoCoord.cpp x2, test_main.cpp x3) to 1-2 lines each in a supplementary commit.

@coderabbitai review

File: test/test_geocoord_distance/test_main.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
+#include "gps/GeoCoord.h"
+#include <cmath>
```
[the remark this answers is not in the document]
Comment (ndoo) [reply test/test_geocoord_distance/test_main.cpp:2]: Fixed in acf4835 — replaced `M_PI` with a local `constexpr double kPi`, qualified `cos`/`sin`/`acos`/`fabs` as `std::`, and added the `<cstdio>` include `snprintf` needs. This also brings the test in line with `GeoCoord.h`'s own existing convention of defining its own `PI` constant rather than relying on `M_PI`.

File: test/test_geocoord_distance/test_main.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+// Pins latLongToMeter()'s equirectangular-approximation accuracy against the original spherical
+// law of cosines, so a future change can't silently regress it.
+
+static double referenceSphericalLawOfCosines(double lat_a, double lng_a, double lat_b, double lng_b)
+{
+    double a1 = lat_a * M_PI / 180.0;
+    double a2 = lng_a * M_PI / 180.0;
+    double b1 = lat_b * M_PI / 180.0;
+    double b2 = lng_b * M_PI / 180.0;
```
[the remark this answers is not in the document]
Comment (ndoo) [reply test/test_geocoord_distance/test_main.cpp]: Fixed in acf4835 — replaced `M_PI` with a local `constexpr double kPi`, qualified `cos`/`sin`/`acos`/`fabs` as `std::`, and added the `<cstdio>` include `snprintf` needs. This also brings the test in line with `GeoCoord.h`'s own existing convention of defining its own `PI` constant rather than relying on `M_PI`.

File: test/test_geocoord_distance/test_main.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+static void assertWithinPercent(double expected, double actual, double pct, const char *msg)
+{
+    if (expected < 1.0)
+        return; // near-zero distances: relative error is meaningless, skip
+    double err = fabs(actual - expected) / expected * 100.0;
+    if (err > pct) {
+        char buf[160];
+        snprintf(buf, sizeof(buf), "%s: expected=%.1f actual=%.1f err=%.2f%% (limit %.2f%%)", msg, expected, actual, err, pct);
```
[the remark this answers is not in the document]
Comment (ndoo) [reply test/test_geocoord_distance/test_main.cpp:49]: Fixed in acf4835 — replaced `M_PI` with a local `constexpr double kPi`, qualified `cos`/`sin`/`acos`/`fabs` as `std::`, and added the `<cstdio>` include `snprintf` needs. This also brings the test in line with `GeoCoord.h`'s own existing convention of defining its own `PI` constant rather than relying on `M_PI`.

File: test/test_geocoord_distance/test_main.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        arg = -1.0;
+    return 6366000 * acos(arg);
+}
+
+static void assertWithinPercent(double expected, double actual, double pct, const char *msg)
+{
+    if (expected < 1.0)
+        return; // near-zero distances: relative error is meaningless, skip
+    double err = fabs(actual - expected) / expected * 100.0;
+    if (err > pct) {
```
[the remark this answers is not in the document]
Comment (ndoo) [reply test/test_geocoord_distance/test_main.cpp:47]: Fixed in acf4835 — near-zero expected distances (<1m) now assert a 0.5m absolute bound instead of skipping validation entirely, so a badly-broken implementation at that scale would still be caught.

File: src/gps/GeoCoord.cpp [hunk 5]
```diff
@@ -433,8 +433,27 @@ void GeoCoord::convertWGS84ToOSGB36(const double lat, const double lon, double &
...
-/// Ported from my old java code, returns distance in meters along the globe
-/// surface (by Haversine formula)
+// Piecewise-linear cos(|latitude|) lookup, denser near 90 degrees where cos() is small and fast-
+// changing. Avoids linking cos()/sin()/acos() (~4.4KB) for a display/movement-threshold utility.
+static double cosLatitudeApprox(double latRad)
+{
+    static constexpr float kLatDeg[] = {0, 15, 30, 45, 55, 65, 72, 78, 82, 85, 87, 88.5, 89.5, 90};
+    static constexpr float kCosVal[] = {1.0f,      0.965926f, 0.866025f, 0.707107f, 0.573576f, 0.422618f, 0.309017f,
+                                        0.207912f, 0.139173f, 0.087156f, 0.052336f, 0.026177f, 0.008727f, 0.0f};
+    constexpr size_t n = sizeof(kLatDeg) / sizeof(kLatDeg[0]);
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/gps/GeoCoord.cpp]: Fixed in acf4835 — added a `static_assert` that `kLatDeg` and `kCosVal` stay the same length, so an edit to one array without the other now fails at compile time instead of risking an out-of-bounds read or silently wrong interpolation.

File: test/test_geocoord_distance/test_main.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    TEST_ASSERT_LESS_THAN_FLOAT(1000000.0f, actual); // sanity: nowhere near the naive-bug's ~40,000km
+}
+
+static void test_symmetry(void)
+{
+    // distance(a,b) should equal distance(b,a)
+    double d1 = GeoCoord::latLongToMeter(51.5074, -0.1278, 48.8566, 2.3522);
+    double d2 = GeoCoord::latLongToMeter(48.8566, 2.3522, 51.5074, -0.1278);
+    TEST_ASSERT_FLOAT_WITHIN(0.01f, d1, d2);
+}
```
[the remark this answers is not in the document]
Comment (ndoo) [reply test/test_geocoord_distance/test_main.cpp:112]: Left this one as-is rather than loosening it - I think 0.01m is actually the right tolerance here, not brittle.

Traced through why: `meanLat = (a1+b1)/2` is commutative (bit-identical regardless of argument order), and `dLng = b2-a2` vs the reverse call's `a2-b2` are exact IEEE-754 negations of each other (floating-point negation is lossless). Since both get squared (`x*x`, `y*y`) before the final `sqrt()`, the sign flip cancels out exactly - the two directions should be bit-identical, not just close, including through the antimeridian wraparound branch (verified that branch preserves the same exact-negation property).

Confirmed empirically: ran `distance(a,b)` vs `distance(b,a)` (with the actual `float` return-type rounding applied) across 5,000+ random point pairs plus the antimeridian/near-pole edge cases specifically, and the max observed difference was exactly 0.0 meters - fully bit-identical every time. So 0.01m is already far looser than the true precision, and tightening my intuition even further wouldn't find a counterexample with this implementation. Loosening it to 0.1m/1m would just make the test less sensitive to a *real* future regression (e.g. someone changing the mean-latitude calc to something non-commutative) without buying any actual robustness today.

Happy to add a comment noting this if it'd help future readers avoid re-litigating the same question.

**Comment by ndoo** [comment]
Had CC address the Copilot reviews @NomDeTom, you may wish to resolve the threads and request a re-review from Copilot or let me know if I should take out of draft.

The changes seem restricted to the unit tests.

@coderabbitai review

Because I think it's worth a 2nd opinion

**Comment by ndoo** [comment]
Taken out of draft but would love a review from someone who actually has some expertise in geospatial coordinate systems.

**Comment by mrpatrick1991** [comment]
I think the idea that we avoid using trig functions to save flash is good, particularly on the STM32WL. The equirectangular coordinates are a good idea to avoid using haversine or something analogous. 

I'm less convinced that this needs to exist for ESP32 and NRF52840.

Why aren't we using one of the standard approximations for cosine? We can avoid the dependency, save the memory, avoid a loop (though, this is probably fast in practice), and be more accurate.

Example: https://www.ganssle.com/approx.htm

**Comment by ndoo** [comment]
> Why aren't we using one of the standard approximations for cosine? We can avoid the dependency, save the memory, avoid a loop (though, this is probably fast in practice), and be more accurate.
> 
> Example: https://www.ganssle.com/approx.htm

Ooh, thanks for the reference. Will implement.

> I'm less convinced that this needs to exist for ESP32 and NRF52840.

@NomDeTom reports flash on NRF52 Promicro to be at/exceeding limits, so we want it there at least. ESP32 probably doesn’t need it, but for parity (and to make space for ideas(tm)), probably worth not gating it out?

**Comment by mrpatrick1991** [comment]
> > Why aren't we using one of the standard approximations for cosine? We can avoid the dependency, save the memory, avoid a loop (though, this is probably fast in practice), and be more accurate.
> > Example: https://www.ganssle.com/approx.htm
> 
> Ooh, thanks for the reference. Will implement.
> 
> > I'm less convinced that this needs to exist for ESP32 and NRF52840.
> 
> @NomDeTom reports flash on NRF52 Promicro to be at/exceeding limits, so we want it there at least. ESP32 probably doesn’t need it, but for parity (and to make space for ideas(tm)), probably worth not gating it out?

If we're against the wall on the NRF52 as well I agree: keeping it uniform across targets is a good thing.

**Comment by ndoo** [comment]
Replaced the lookup table with the `cos_52` minimax polynomial from that article: 4 coefficients, no loop, no range reduction (latitude in radians is already within `[-π/2, π/2]`).

| | develop (no PR) | This PR, lookup table | This PR, `cos_52` |
|---|---|---|---|
| Max abs error vs `cos()`, 0–90° | 0 (exact) | 0.0085 | 0.0000067 |
| Flash, `wio-e5` | 233,776 B | 228,000 B (-5,776) | 227,776 B (-6,000) |
| Flash, `seeed_wio_tracker_L1` | 751,224 B | 750,120 B (-1,104) | 749,896 B (-1,328) |

Since this is smaller and more accurate than the table on every board, there's no accuracy/flash tradeoff left to weigh per-platform.

Pushed as [846375a](https://github.com/meshtastic/firmware/pull/11184/commits/846375ade).

**Comment by ndoo** [comment]
@coderabbitai review

**Comment by ndoo** [comment]
> > > Why aren't we using one of the standard approximations for cosine? We can avoid the dependency, save the memory, avoid a loop (though, this is probably fast in practice), and be more accurate.
> > > Example: https://www.ganssle.com/approx.htm
> > 
> > 
> > Ooh, thanks for the reference. Will implement.
> > > I'm less convinced that this needs to exist for ESP32 and NRF52840.
> > 
> > 
> > @NomDeTom reports flash on NRF52 Promicro to be at/exceeding limits, so we want it there at least. ESP32 probably doesn’t need it, but for parity (and to make space for ideas(tm)), probably worth not gating it out?
> 
> If we're against the wall on the NRF52 as well I agree: keeping it uniform across targets is a good thing.

How does the current approach look?

**Comment by mrpatrick1991** [comment]
Can you re-run the error analysis above using the new approximation? If it's good, I think we should keep it for the 32WL behind an #ifdef or something and leave it out for the other platforms.

**Comment by ndoo** [comment]
> Can you re-run the error analysis above using the new approximation? If it's good, I think we should keep it for the 32WL behind an #ifdef or something and leave it out for the other platforms.

PR text has been updated. I believe we want to at least scope STM32WL + NRF52 behind the ifdef if we really do not want ESP platforms covered; @NomDeTom

**Comment by ndoo** [comment]
Anything pending here? Would love to get some flash savings on STM32WL (would be happy to gate the PR to STM32WL if the regression concerns are the main issue)

**Comment by ndoo** [comment]
Added the possibility to use a macro `MESHTASTIC_TRIG_APPROX=0` to restore the old behavior for edge-case users who need the old behavior, i.e. expecting to pass close to/over Earth's poles.

Edit: The diff also looks a bit cleaner now that there are no deletions to the old code path.

Took the opportunity to rebase the PR commits on develop HEAD.

**Review by caveman99 — APPROVED** [review]
